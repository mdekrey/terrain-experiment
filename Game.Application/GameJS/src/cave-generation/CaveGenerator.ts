import {
  GameCoordinates
} from "../game/GameCoordinates";
import { ajax } from "rxjs/ajax";

export interface Cave {
  offset: GameCoordinates;
  isSolid: boolean[][];
  treasure: GameCoordinates[];
  entrance: GameCoordinates;
}

export class CaveGenerator {
  private readonly result: Promise<{
    isSolid: boolean[][];
    entrance: GameCoordinates;
    treasure: GameCoordinates[];
  }>;
  private readonly position: GameCoordinates;
  private readonly offset: GameCoordinates;

  constructor(
    position: GameCoordinates,
    offset: GameCoordinates
  ) {
    this.position = position;
    this.offset = offset;

    this.result = new Promise(this.buildMap);
  }

  private readonly buildMap = async (
    resolve: (t: { isSolid: boolean[][]; treasure: GameCoordinates[], entrance: GameCoordinates }) => void
  ) => {
    const response = await ajax({ url: "/api/terrain/cave", body: { Coordinates: { X: Math.round(this.position.x), Y: Math.round(this.position.y) } }, method: "POST", headers: { 'Content-Type': 'application/json' } })
      .toPromise();
    const result = response.response as { map: boolean[][]; treasure: GameCoordinates[], entrance: GameCoordinates };
    resolve({ isSolid: result.map, treasure: result.treasure, entrance: result.entrance });
  };

  get cave(): Promise<Cave> {
    return this.result.then(cave => ({
      isSolid: cave.isSolid,
      entrance: cave.entrance,
      treasure: cave.treasure,
      offset: this.offset
    }));
  }
}
