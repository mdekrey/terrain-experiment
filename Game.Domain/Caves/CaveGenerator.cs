using LibNoise;
using LibNoise.Generator;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Game.Domain.Caves
{
    public class CaveGenerator
    {
        const int deathLimit = 3;
        const int birthLimit = 4;
        const int treasureLimit = 5;

        private readonly int width;
        private readonly int height;
        private readonly int normalizationRounds;
        private readonly ModuleBase perlin;
        private bool[][] map;
        private readonly Task<Cave> building;



        bool Random(int x, int y) => perlin.GetValue(x / 3f, y / 3f, 0) < 0;

        public CaveGenerator(int width, int height, int normalizationRounds, int seed)
        {
            this.width = width;
            this.height = height;
            this.normalizationRounds = normalizationRounds;

            this.perlin = new Perlin(lacunarity: 3.4f, seed: seed);
            this.map = (from y in Enumerable.Range(0, height)
                        select (from x in Enumerable.Range(0, width)
                                select Random(x, y)).ToArray()).ToArray();
            building = BuildMap();
        }

        private int CountAliveNeighbours(CaveCoordinate position)
        {
            var count = 0;
            foreach (var dir in CaveCoordinate.EightDirections)
            {
                var nb = dir + position;
                //If it's at the edges, consider it to be REALLY solid
                if (
                  nb.x < 0 ||
                  nb.y < 0 ||
                  nb.x >= width ||
                  nb.y >= height
                )
                {
                    count = count + 2;
                }
                else if (this.map[nb.y][nb.x])
                {
                    count = count + 1;
                }
            }
            return count;
        }

        private async Task<Cave> BuildMap()
        {
            await Task.Yield();
            for (var count = 0; count < this.normalizationRounds; count++)
            {
                this.DoSimulationStep();
            }

            this.FillSmallerAreas();
            this.RemoveSingleY();

            var treasure = FindTreasure();
            return new Cave
            {
                Map = map,
                Entrance = treasure[0],
                Treasure = treasure.Skip(1).ToArray(),
            };
        }

        private CaveCoordinate[] FindTreasure()
        {
            return (from y in this.map.Select((_, y) => y)
                    from v in this.map[y].Select((currentCell, x) => (currentCell, x))
                    let currentCell = v.currentCell
                    where !currentCell
                    let x = v.x
                    let r = new CaveCoordinate(x, y)
                    let nbs = this.CountAliveNeighbours(r)
                    where nbs >= treasureLimit
                    select r).ToArray();
        }

        private void FillSmallerAreas()
        {
            var areas = new Dictionary<string, int>();
            string[][] currentAreas = this.map.Select(_ => new string[_.Length]).ToArray();

            var threshold = (this.width * this.height) / 4;
            string shortcutSymbol = null;
            for (var y = 0; y < this.map.Length && shortcutSymbol == null; y++)
            {
                for (var x = 0; x < this.map[y].Length && shortcutSymbol == null; x++)
                {
                    if (!this.map[y][x] && currentAreas[y][x] == null)
                    {
                        var (symbol, count) = this.FloodFill(
                          new CaveCoordinate(x, y),
                          currentAreas,
                          areas
                        );
                        if (count > threshold)
                        {
                            shortcutSymbol = symbol;
                        }
                    }
                }
            }
            var topSymbol =
              shortcutSymbol ?? areas.OrderByDescending(v => v.Value).First().Key;

            for (var y = 0; y < this.map.Length; y++)
            {
                for (var x = 0; x < this.map[y].Length; x++)
                {
                    if (currentAreas[y][x] != topSymbol)
                    {
                        this.map[y][x] = true;
                    }
                }
            }
        }

        private (string symbol, int count) FloodFill(CaveCoordinate start, string[][] currentAreas, Dictionary<string, int> areas)
        {
            var symbol = $"{start.x}x{start.y}";
            var count = 0;
            var queue = new Queue<CaveCoordinate>();
            queue.Enqueue(start);
            while (queue.Count > 0)
            {
                var n = queue.Dequeue();
                foreach (var dir in CaveCoordinate.FourDirections)
                {
                    var r = (n + dir);
                    if (!this.map[r.y][r.x] && currentAreas[r.y][r.x] == null)
                    {
                        count++;
                        currentAreas[r.y][r.x] = symbol;
                        queue.Enqueue(r);
                    }
                }
            }
            areas[symbol] = count;
            return (symbol, count);
        }

        void RemoveSingleY()
        {
            for (var y = 1; y < this.map.Length - 1; y++)
            {
                for (var x = 0; x < this.map[y].Length; x++)
                {
                    if (
                      this.map[y][x] &&
                      !this.map[y + 1][x] &&
                      !this.map[y - 1][x]
                    )
                    {
                        this.map[y][x] = false;
                    }
                }
            }
        }

        private void DoSimulationStep()
        {
            this.map = this.map.Select((_, y) => _.Select((currentCell, x) =>
            {
                var nbs = CountAliveNeighbours(new CaveCoordinate(x, y));
                if (currentCell)
                {
                    return nbs >= deathLimit;
                }
                else
                {
                    return nbs > birthLimit;
                }
            }).ToArray()).ToArray();
        }

        public Task<Cave> Generate()
        {
            return building;
        }
    }
}
