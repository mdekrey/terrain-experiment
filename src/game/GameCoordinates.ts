export interface GameCoordinates {
    readonly x: number;
    readonly y: number;
}

export function interpolate(factor: number, { x: x1, y: y1 }: GameCoordinates, { x: x2, y: y2 }: GameCoordinates) {
    return { x: factor * (x2 - x1) + x1, y: factor * (y2 - y1) + y1  };
}