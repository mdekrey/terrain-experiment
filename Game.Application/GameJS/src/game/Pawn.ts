import { GameCoordinates, interpolate } from "./GameCoordinates";
import { Interpolated, makeInterpolated, isInterpolated } from "./Interpolated";
import { Direction } from "./Direction";
import { PawnType } from "./PawnType";

export class Pawn {

    private _position: GameCoordinates | Interpolated<GameCoordinates> = { x: 0, y: 0};

    facing: Direction = Direction.Down;
    type: PawnType = PawnType.Hero;


    moveTo(target: GameCoordinates, facing: Direction, delay: number = 0) {
        const now = Date.now();
        this._position = delay
            ? makeInterpolated(this.position(), target, now, now + delay, interpolate)
            : target;
        this.facing = facing;
    }

    isDoneMoving() {
        return isInterpolated(this._position) ? this._position.isComplete : true;
    }

    readonly position = (): GameCoordinates => {
        return isInterpolated(this._position) ? this._position.value() : this._position;
    }
}
