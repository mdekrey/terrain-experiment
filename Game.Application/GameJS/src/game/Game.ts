import { TerrainSettings } from "../terrain-generation/TerrainSettings";
import { TerrainGenerator, TerrainCache, VisualTerrainType } from "../terrain-generation";
import { Pawn } from "./Pawn";
import { Cave, CaveGenerator } from "../cave-generation";
import { BehaviorSubject } from "rxjs";
import { addCoordinates, GameCoordinates } from "./GameCoordinates";
import { Direction } from "./Direction";
import { PawnType } from "./PawnType";

export interface OverworldGameMode {
    mode: "Overworld";
}

export interface DetailGameMode {
    mode: "Detail";
}

export interface CaveGameMode {
    mode: "Cave";
    cave: Cave;
}

export interface LoadingGameMode {
    mode: "Loading";
}

export type GameMode = OverworldGameMode | CaveGameMode | LoadingGameMode | DetailGameMode;

const zoomExp = 2;

export class Game {
    readonly terrain: TerrainCache;
    readonly playerPawn: Pawn;
    readonly gameMode$ = new BehaviorSubject<GameMode>({ mode: "Overworld" })
    readonly otherPlayers: Pawn[];

    readonly overworldZoom = Math.pow(10, zoomExp) * 4;
    readonly localZoom = Math.pow(10, zoomExp + 2) * 4;

    constructor(settings: TerrainSettings, playerPawn: Pawn) {
        const terrainGenerator = new TerrainGenerator(settings);
        this.terrain = new TerrainCache(terrainGenerator);
        this.playerPawn = playerPawn;
        if (process.env.NODE_ENV === "development") {
            (window as any).game = this;
        }

        const types = Object.values(PawnType);
        const generatePlayer = () => {
            const result = new Pawn();
            result.moveTo({ x: Math.floor((Math.random() - 0.5) * 50) / this.overworldZoom, y: Math.floor((Math.random() - 0.5) * 50) / this.overworldZoom }, Math.floor(Math.random() * 4));
            result.type = types[Math.floor(Math.random() * types.length)];
            return result;
        }
        this.otherPlayers = [generatePlayer(), generatePlayer(), generatePlayer(), generatePlayer()];
    }

    async enterDetail() {
        if (this.playerPawn.isDoneMoving()) {
            const position = this.playerPawn.position();
            if (this.terrain.getAt(position.x, position.y, false).indexOf("Cave") !== -1) {
                this.enterCave();
            } else {
                this.gameMode$.next({ mode: "Detail" });
            }
        }
    }

    async enterCave() {
        if (this.playerPawn.isDoneMoving()) {
            const position = this.playerPawn.position();
            this.gameMode$.next({ mode: "Loading" });

            const gen = new CaveGenerator({ x: position.x * this.overworldZoom, y: position.y * this.overworldZoom }, addCoordinates(position, { x: -0.5 / this.overworldZoom, y: -0.5 / this.overworldZoom }));
            const cave = await gen.cave;
            this.playerPawn.moveTo(addCoordinates(cave.offset, { x: cave.entrance.x / this.localZoom, y: cave.entrance.y / this.localZoom }), Direction.Down);
            this.gameMode$.next({ mode: "Cave", cave });
        }
    }

    moveToOverworld() {
        const gameMode = this.gameMode$.value;
        if (this.playerPawn.isDoneMoving()) {
            const position = this.playerPawn.position();
            if (gameMode.mode === "Cave") {
                const { offset, entrance } = gameMode.cave;
                const x = Math.round((position.x - offset.x) * this.localZoom);
                const y = Math.round((position.y - offset.y) * this.localZoom);
                if (x !== entrance.x || y !== entrance.y) {
                    console.log("not at entrance", x, y, entrance);
                    return;
                }
            }
            const targetPosition = { x: Math.round(position.x * this.overworldZoom) / this.overworldZoom, y: Math.round(position.y * this.overworldZoom) / this.overworldZoom }
            if (this.isOpenSpace(targetPosition, true)) {
                this.playerPawn.moveTo(targetPosition, Direction.Down);
                this.gameMode$.next({ mode: "Overworld" });
            } else {
                console.log("not in open space", targetPosition)
            }
        }
    }

    movePlayerTo(worldCoordinate: GameCoordinates, facing: Direction) {
        if (!this.playerPawn.isDoneMoving()) {
            return;
        }
        if (this.isOpenSpace(worldCoordinate)) {
            this.playerPawn.moveTo(worldCoordinate, facing, 100);
        } else {
            this.playerPawn.facing = facing;
        }
    }

    isOpenSpace(worldCoordinate: GameCoordinates, forceOverworld = false) {
        const gameMode = this.gameMode$.value;

        const overworldCheck = () => {
            const category = this.terrain.getAt(worldCoordinate.x, worldCoordinate.y, false);
            return !category.some(t => !isPassable(t));
        };
        if (forceOverworld) {
            return overworldCheck();
        }
        switch (gameMode.mode) {
            case "Cave":
                const {cave} = gameMode;
                const caveY = Math.round((worldCoordinate.y - cave.offset.y) * this.localZoom);
                const caveX = Math.round((worldCoordinate.x - cave.offset.x) * this.localZoom);
                return !cave.isSolid[caveY][caveX];
            case "Loading":
                return false;
            case "Overworld":
                {
                    return overworldCheck();
                }
            case "Detail":
                {
                    const category = this.terrain.getAt(worldCoordinate.x, worldCoordinate.y, true);
                    return !category.some(t => !isPassable(t));
                }
        }
        function isPassable(category: VisualTerrainType) {
            return category !== "SnowyMountains" && category !== "Mountains" && category !== "ShallowWater" && category !== "DeepWater";
        }
    }
}
