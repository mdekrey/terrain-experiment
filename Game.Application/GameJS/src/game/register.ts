import { injectorBuilder, Scope } from "../injector";
import { TerrainCache } from "../terrain-generation";
import { Pawn } from "./Pawn";
import { Game } from "./Game";
import "../api";

declare module "../injector/InjectedServices" {
  interface InjectedServices {
    terrainCache: TerrainCache;
    game: Game;
    player: Pawn;
  }
}

injectorBuilder.set(
    "player",
    Scope.Component,
    () => new Pawn()
);

injectorBuilder.set("game", Scope.Component, resolver => new Game(resolver("player"), resolver("terrainService")));
injectorBuilder.set(
  "terrainCache",
  Scope.Component,
  resolver => resolver("game").terrain
);
