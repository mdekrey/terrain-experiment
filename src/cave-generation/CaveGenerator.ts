import { Perlin } from "libnoise-ts/module/generator";
import { GameCoordinates, eightDirections, addCoordinates, fourDirections } from "../game/GameCoordinates";

const deathLimit = 3;
const birthLimit = 4;
const treasureLimit = 5;

interface Cave {
    isSolid: boolean[][];
    treasure: GameCoordinates[];
}

export class CaveGenerator {
    private readonly perlin = new Perlin();
    private readonly result: Promise<Cave>;

    constructor(seed: number, width: number, height: number, normalizationRounds: number) {
        this.perlin.seed = seed;
        this.perlin.lacunarity = 3.4;

        const random = (x: number, y: number) => this.perlin.getValue(x, y, 0) < -0.05;

        this.result = new Promise<Cave>((resolve, reject) => {
            let map = Array.from(Array(height).keys()).map((_, y) => Array.from(Array(width).keys()).map((_, x) => random(x, y)));
            let count = 0;
            let filled = false;

            function countAliveNeighbours(position: GameCoordinates){
                let count = 0;
                for (const dir of eightDirections) {
                    const {x: nb_x, y: nb_y} = addCoordinates(dir, position);
                    //If it's at the edges, consider it to be REALLY solid
                    if(nb_x < 0 || nb_y < 0 ||
                            nb_x >= map.length ||
                            nb_y >= map[0].length){
                        count = count + 2;
                    }
                    else if(map[nb_y][nb_x]){
                        count = count + 1;
                    }
                }
                return count;
            }

            function doSimulationStep() {
                map = map.map((_, y) => _.map((currentCell, x) => {
                    const nbs = countAliveNeighbours({x, y});
                    if (currentCell) {
                        return nbs >= deathLimit;
                    } else {
                        return nbs > birthLimit;
                    }
                }));
            }

            function findTreasure() {
                return map.map((_, y) => _.map((currentCell, x): (false | GameCoordinates) => {
                    const nbs = countAliveNeighbours({x, y});
                    return !currentCell && nbs >= treasureLimit ? { x, y } : false;
                })).reduce((prev, next) => [...prev, ...next], []).filter((v): v is GameCoordinates => Boolean(v));
            }

            function fillSmallerAreas() {
                const areas = new Map<Symbol, number>();
                const currentAreas: Symbol[][] = map.map(_ => []);

                function floodFill(startX: number, startY: number) {
                    const symbol = Symbol();
                    let count = 0;
                    const queue = [{x: startX, y: startY}];
                    let n: GameCoordinates | undefined;
                    while (n = queue.pop()) {
                        for (const dir of fourDirections) {
                            const {x, y} = addCoordinates(n, dir);
                            if (!map[y][x] && !currentAreas[y][x]) {
                                count++;
                                currentAreas[y][x] = symbol;
                                queue.push({ x, y });
                            }
                        }
                    }
                    areas.set(symbol, count);
                }

                for (let y = 0; y < map.length; y++) {
                    for (let x = 0; x < map[y].length; x++) {
                        if (!map[y][x] && !currentAreas[y][x]) {
                            floodFill(x, y);
                        }
                    }
                }
                const topSymbol = Array.from(areas.keys()).sort((a, b) => areas.get(b)! - areas.get(a)!)[0];

                for (let y = 0; y < map.length; y++) {
                    for (let x = 0; x < map[y].length; x++) {
                        if (currentAreas[y][x] !== topSymbol) {
                            map[y][x] = true;
                        }
                    }
                }
            }

            function step() {
                if (count++ < normalizationRounds) {
                    doSimulationStep();
                } else if (!filled) {
                    fillSmallerAreas();
                    filled = true;
                } else {
                    resolve({ isSolid: map, treasure: findTreasure() });
                    return;
                }
                setTimeout(step);
            }

            setTimeout(step)
        });
    }

    get map() {
        return this.result.then(cave => cave.isSolid);
    }

    get entrance() {
        return this.result.then(cave => cave.treasure[0]);
    }

    get treasure() {
        return this.result.then(cave => cave.treasure.slice(1));
    }
}
