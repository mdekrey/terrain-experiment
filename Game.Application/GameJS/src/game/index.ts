import "./register";
export * from "./Pawn";
export * from "./Direction";
export * from "./PawnType";
export * from "./Game";
export * from "./GameCoordinates";
export * from "./PawnType";
