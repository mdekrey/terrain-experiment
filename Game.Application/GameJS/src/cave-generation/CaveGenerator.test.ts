import { CaveGenerator } from "./CaveGenerator";
import { coordinatesEqual } from "../game/GameCoordinates";

it.skip("generates some terrain", async () => {
  const caveGenerator = new CaveGenerator(0, 50, 50, 2, {x: 0, y: 0});

  const cave = await caveGenerator.cave;
  const map = cave.isSolid;
  const treasure = cave.treasure;
  const entrance = cave.entrance;
  const mapDisplay = map
    .map((row, y) =>
      row
        .map((e, x) =>
            coordinatesEqual({ x, y }, entrance) ? "/"
                : treasure.find(entry => coordinatesEqual({x, y}, entry)) ? "*"
                : e ? "â–ˆ"
                : " "
        )
        .join("")
    )
    .join("\n");

  expect(mapDisplay).toMatchSnapshot();
});
