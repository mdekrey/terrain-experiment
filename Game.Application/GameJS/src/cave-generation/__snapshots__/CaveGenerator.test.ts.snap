// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generates a cave at seed 0 1`] = `
"██████████████████████████████████████████████████
██████████     ███████████       █████████████████
█████████/      █████████         ████████████████
██████████      ████████          ████████████████
███████████    ██*   ███* █████  █████████████████
████████████   ██     █████████  █████████████████
████████████          ████████    ████████████████
█████████████    ███  ███████     ████████████████
█████████████    ████  █████     *████████████████
████████████      ███          ███████████████████
████████████      ████        ████████████████*███
████████████      █████      ████████████████   *█
████████████       ████████  ████████████████    █
█████████████      ████████  ████████████████    █
█████████████       ███████  ███████████████     █
████████████         ██████  █████████████       █
█████████*    ███     █████ █████████████       ██
██████████   █████     ███  █████████████       ██
███████████████████         ██  ███████          █
█████████████████████            █████       ██ *█
██████████████████████           █████       █████
████████████████████████    ██     ████      █████
█████████*  *█████████████████       ███████ █████
█████████    █████████████████        ██████  ████
█████████    ██    ██    ██████               ████
█████████   ██            █████*█████         ████
█████████   ██             ████ █████   ██     ███
█████████             ██    ██   ███    █████  *██
██████████    ███████████   ██   ███    ██████*███
██████████   ██████████████      ██     ██████████
█████████    ████████    ██             ██████████
█████████       ████               ██  ███████████
█████████*      *██         ██*██  ██  ███████████
████████████  ████          █████      ███████████
██████████████████      ██  █████     ████████████
██████████████████     ████  ████    █████████████
███████████████████   █████   ████  ██████████████
████████████████████████████  ████████████████████
████████████████████████████  ████████████████████
████████████████████████████   ███████████████████
█████████████████████████████ *███████████████████
██████████████████████████████████████████████████
██████████████████████████████████████████████████
██████████████████████████████████████████████████
██████████████████████████████████████████████████
██████████████████████████████████████████████████
██████████████████████████████████████████████████
██████████████████████████████████████████████████
██████████████████████████████████████████████████
██████████████████████████████████████████████████"
`;

exports[`generates a cave at seed 38361 1`] = `
"██████████████████████████████████████████████████
████████████████████         █████ /██████████████
███████████████████           ███   ██████████████
██████████████████             █     ██  █████████
██████████████████   ████        ██       ████████
██████████████████   ████   ██   ███      ████████
██████████████████* ████    ██    █████    ███████
████████████████████████   ████   ███████ *███████
███████████████████████  ██████   ███*████████████
███████████     ██████*  ███      ██   ███████████
██████████          ███*███       ██   ███████████
██████████           ██████       ██  ████████████
█████████    ██       █████       ██  ████████████
████████   ████   █              ███  ████████████
███████    ████   ██            ███    ███████████
██████    ████    ██████    █  ██       ██████████
██████   █████    ███████  █████        ██████████
█████   ██████    ███████  █████         █████████
██*     ███  ██  *██████    █████    ██   ████████
██            █████████      ████  █████  ████████
███           ████████       ███  ██████  ████████
███    ██       ██       ██       ██████     █████
██   ████  ██            ████    ██████       ████
█    ████  ████    ██    ████   █████           *█
█      ██   ███    ███         █████       ████ *█
██          ███   ███████      █████    ██████████
███         ███   ████████      ████    ██████████
█████        ██   ████████████   ███    ███   ████
█████   ██   ██   █████████████  ███   ███      *█
█████   ███       ████████*████  ███   ███       █
██████   ███       ██████   ██    ██    █     █*██
███████   ██* ██     ████   ██    ███      ███████
███████   ██████      ███   ██    ███      ███████
███████   █████       ██    ██     █      *███████
███████   █████* ██   ██   ████        ███████████
███████   █████████        █████       ███████████
██  ███* ██████████       ███████     ████████████
█    ██████████████      █████████████████████████
█*   █████  ████████     █████████████████████████
███    ██    ████████    █████    ██    ██████████
████         █████████  ███              █████████
█████        █████████  ██           █   █████████
██████       █████  ██      ██      ███    ███████
█████████                  ███      ███    *██████
█████████                ███*      ███     ███████
██████████      ████     ███*     ████* ██████████
███████████████████████  █████████████████████████
███████████████████████   ████████████████████████
███████████████████████* *████████████████████████
██████████████████████████████████████████████████"
`;

exports[`generates a cave at seed 66335 1`] = `
"██████████████████████████████████████████████████
██████████████████████████████████████████████████
██████████████████████████████████████████████████
██████████████████████████████████████████████████
██████████████████████████████████████████████████
██████████████████████████████████████████████████
██████████████████████████████████████████████████
███████████    ███████████████████████████████████
██████████      ██████████████████████████████████
█████████          ███████████████████████████████
████████    ██      ██████████████████████████████
███████     ███      █████████████████████████████
██████      ██████    ████████████████████████████
██████  ██  ███████   ████████████████████████████
██████/███  ███████   ████████████████████████████
██████*███   █████    ████████████████████████████
██*    ██     ███     ████████████████████████████
██     ██*     ██     *███████████████████████████
███    █████   ███    ████████████████████████████
███   ███████  ███████████████████████████████████
███   ███████  ███████████████████████████████████
██    ██████    ██████████████████████████████████
█*   ███████    ██████████████████████████████████
██   ██████    ███████████████████████████████████
███  █       █████████████████████████████████████
███         ██████████████████████████████████████
███         ██████████████████████████████████████
███    ██   ██████████████████████████████████████
████  ████*███████████████████████████████████████
████  ████████*███████████████████████████████████
████   ██████   ██████████████████████████████████
████   ██████   ██████████████████████████████████
████    *████   ██████████████████████████████████
█*       ███    ██████████████████████████████████
█        ███    ████████   ███████████████████████
█        ███    ███████     ██████████████████████
██  ██   ███    █████         ████████████████████
███████  ███    ████           ███████████████████
███████  ███     ███           ███████*███████████
███████   ██          █████       ███   ██████████
███████        ██    ██████        ██   ██████████
████████       ██    ██████        ██   ██████████
█████████*██    █      ████             ██████████
████████████    █       ████  █████     ██████████
████████████    ██       ███  ██████    ██████████
█████████████  ██████    ███   █████    ██████████
██████████████████████   ████          ███████████
██████████████████████    ███        █████████████
███████████████████████  *███*    ████████████████
██████████████████████████████████████████████████"
`;
